### ISSUE TITLE
TPRegion should output 2 new fields: active & predicted columns separately

### ISSUE THREAD

**Issue by breznak** [numenta/nupic#2081]
Depends on https://github.com/numenta/nupic/issues/1450

It will be useful if the output of the TP region provided separate fields for the `active` & `predicted` columns; 

currently these are combined to the default output as defined by the TP outputType - either of them, or both. 

Use case: AnomalyRegion could depend only on TPRegion (and not SP, which is now required to get the active columns, [see](https://github.com/numenta/nupic/blob/master/examples/network/network_api_demo.py#L158) ), I wonder if the [lrnActiveCells](https://github.com/numenta/nupic/blob/master/nupic/regions/TPRegion.py#L576) is the same as activeOut suggested for this usecase (maybe not, as it's only used in `learn` phase)

TODO:
- [ ] TPRegion.output provides new activeOut and predictedOut fields
- [ ] also we should remove [anomalyScore](https://github.com/numenta/nupic/blob/master/nupic/regions/TPRegion.py#L568) from TPRegion, SPRegion output, and use AnomalyRegion only

**Comment by breznak** [comment]
CC @scottpurdy

**Comment by scottpurdy** [comment]
TPRegion `topDownOut` gives you the predicted columns and the active columns are the input to the TPRegion. If the input comes from the SPRegion then the SPRegion `bottomUpOut` gives you the active columns.

**Comment by scottpurdy** [comment]
FWIW, I think it would be much easier to understand if we gave more explicit names to the inputs and outputs of the regions but need to make the serialization switch before we can change them.

**Comment by breznak** [comment]
> TPRegion topDownOut gives you the predicted columns and the active columns are the input to the TPRegion. 

yes, my point was that if TP output provided separately activeOut and predictedOut, eg Anomalyregion could link just to Tp.output as its input, instead of 2 inputs (where 1 is always TP, other depends if SP/encoder)

**Comment by scottpurdy** [comment]
@breznak - wouldn't it still be two links?

**Comment by breznak** [comment]
but you don't have to care for the underlying layer, see the `if-else` in https://github.com/numenta/nupic/pull/2082#discussion-diff-29110675R1139

**Comment by scottpurdy** [comment]
So to be clear, this is a proposal to make the TM provide its input (active columns) as an output? That would add overhead. It may be useful as you say to avoid having to know what feeds into the TM but I'm not sure how much of a benefit that is practice.

I am pretty strongly in favor of changing the names to be more descriptive though.

**Comment by breznak** [comment]
I was talking about TP, but suppose same holds for TM. 

> ... provide its input (active columns) as an output? 

yes

I'll leave the judgement `cleaner-vs-overhead` up to you...

**Comment by scottpurdy** [comment]
@breznak - Ok, yeah I think I am fine with it, especially if we can avoid extra work when there are no links connected to the active columns output, but seems very low priority, especially compared to getting serialization stuff done so we can rename some of these links and such to be more descriptive.

Are you ok if we update this issue to be specifically for adding the activeColumns output to TPRegion since the predicted columns is already there (just needs a rename)?

**Comment by scottpurdy** [comment]
also just fyi, the anomaly score in the SP is different than the anomaly score in AnomalyRegion or TPRegion. The SP anomaly score is for spatial anomalies, which we experimented with but don't have any current uses of AFAIK

**Comment by rhyolight** [comment]
@breznak said he's AFK for a week. We might not get an answer from him. In the meantime, I want to link this issue to #2094.

**Comment by breznak** [comment]
Reopening with a few comments to the merged PR, thanks. 

https://github.com/numenta/nupic/pull/2206#discussion_r33442075
https://github.com/numenta/nupic/pull/2206#discussion-diff-33441996

**Comment by rhyolight** [comment]
What's the status of this ticket, @breznak ?

**Comment by breznak** [comment]
> What's the status of this ticket

@rhyolight I don't have much time to look at it now, will try to review it.

**Comment by andrewmalta13** [comment]
Is there still work to be done here? @breznak

**Comment by breznak** [comment]
@andrewmalta13 I havent followed the issue for a while, sorry. We'll have to look

**Comment by andrewmalta13** [comment]
ok I will take a look

**Comment by andrewmalta13** [comment]
First comment seems to be addressed by this commit https://github.com/numenta/nupic/commit/6c08555766b20d14b559090d13d08980b67c5d0e

second comment was the naming convention that didn't end in agreement:
predictedActiveCells vs. predictedCells vs. successfullyPredictedCells

Unless I missed something I think this issue should be closed. @breznak  @scottpurdy
